circuit_knitting/utils/transforms.py

```diff
@@ -191,44 +214,60 @@ def _circuit_from_instructions(
 
 def _qubit_map_from_partition_labels(
     partition_labels: Sequence[Hashable],
-) -> tuple[list[tuple[Hashable, int]], dict[Hashable, list[int]]]:
+) -> tuple[list[tuple[Hashable, int] | tuple[None, None]], dict[Hashable, list[int]]]:
     """Generate a qubit map given a qubit partitioning."""
-    qubit_map: list[tuple[Hashable, int]] = []
+    qubit_map: list[tuple[Hashable, int] | tuple[None, None]] = []
     qubits_by_subsystem: MutableMapping[Hashable, list[int]] = defaultdict(list)
     for i, qubit_label in enumerate(partition_labels):
-        current_label_qubits = qubits_by_subsystem[qubit_label]
-        qubit_map.append((qubit_label, len(current_label_qubits)))
-        current_label_qubits.append(i)
+        if qubit_label is None:
+            qubit_map.append((None, None))
+        else:
+            current_label_qubits = qubits_by_subsystem[qubit_label]
+            qubit_map.append((qubit_label, len(current_label_qubits)))
+            current_label_qubits.append(i)
     return qubit_map, dict(qubits_by_subsystem)
 
 
 def _separate_instructions_by_partition(
     circuit: QuantumCircuit,
-    qubit_map: Sequence[tuple[Hashable, int]],
+    qubit_map: Sequence[tuple[Hashable, int] | tuple[None, None]],
 ) -> dict[Hashable, list[int]]:
     """Generate a list of instructions for each partition of the circuit."""
-    unique_labels = unique_by_eq(label for label, _ in qubit_map)
+    unique_labels = unique_by_eq(label for label, _ in qubit_map if label is not None)
     subcircuit_data_ids: dict[Hashable, list[int]] = {
         label: [] for label in unique_labels
     }
 
-    for i, gate in enumerate(circuit.data):
-        partitions_spanned = {
-            qubit_map[circuit.find_bit(qubit).index][0] for qubit in gate.qubits
-        }
-        # All qubits for any gate should belong to one subset
+    for i, inst in enumerate(circuit.data):
+        # Collect the partition labels spanned by the instruction
+        partitions_spanned = set()
+        for qubit in inst.qubits:
+            j = circuit.find_bit(qubit).index
+            partition_id = qubit_map[j][0]
+            if partition_id is None:
+                raise ValueError(
```

Comment:
Do you raise here because `None` is technically `Hashable`? Users could pass in a `Sequence[None]`, and it would obey the `Sequence[Hashable]` type hint, but be an invalid input?

Reply:
> Do you raise here because `None` is technically `Hashable`? Users could pass in a `Sequence[None]`, and it would obey the `Sequence[Hashable]` type hint, but be an invalid input?

Not quite; it's because a label of `None` means "this qubit should be idle," but if this code path is triggered, it's because there is an instruction that uses the qubit in the circuit (i.e., it's not _actually_ idle).

I wonder if the error message (on the following lines) could be clarified in any way.

Reply:
No, I think it's a good message. This arises when a user passes `None` as a label for an active qubit, and I think that's clear here.

I guess what I was getting at is that the user passing `None` on an active qubit accidentally doesn't disobey the type hint, so raising a `ValueError` is appropriate.